lxd/storage_zfs.go

```diff
@@ -2844,107 +2905,411 @@ func (s *storageZfs) StoragePoolResources() (*api.ResourcesStoragePool, error) {
+		var err error
+
+		if s.pool.Config["zfs.clone_copy"] != "" && !shared.IsTrue(s.pool.Config["zfs.clone_copy"]) {
+			err = s.copyVolumeWithoutSnapshotsFull(source)
+		} else {
+			// TODO (monstermunchkin): implement
+			err = s.copyVolumeWithoutSnapshotsSparse(source)
+		}
+		if err != nil {
+			return err
+		}
+	} else {
+		targetStorageVolumeMountPoint := getStoragePoolVolumeMountPoint(s.pool.Name, s.volume.Name)
+
 		poolName := s.getOnDiskPoolName()
-		snapUUID := fmt.Sprintf("copy-%s", uuid.NewRandom().String())
 
-		// create snapshot for zfs to clone from
-		err := zfsPoolVolumeSnapshotCreate(poolName, storagePoolVolumeTypeNameCustom, snapUUID)
+		tmpSnapshotName := fmt.Sprintf("copy-send-%s", uuid.NewRandom().String())
+		err := zfsPoolVolumeSnapshotCreate(poolName, fmt.Sprintf("custom/%s", source.Name), tmpSnapshotName)
+		if err != nil {
+			return err
+		}
+
+		defer zfsPoolVolumeSnapshotDestroy(poolName, fmt.Sprintf("custom/%s", source.Name), tmpSnapshotName)
+
+		currentSnapshotDataset := fmt.Sprintf("%s/custom/%s@%s", poolName, source.Name, tmpSnapshotName)
+		args := []string{"send", "-R", currentSnapshotDataset}
+
+		zfsSendCmd := exec.Command("zfs", args...)
+		targetDataset := fmt.Sprintf("%s/custom/%s", poolName, s.volume.Name)
+		zfsRecvCmd := exec.Command("zfs", "receive", "-F", targetDataset)
+
+		zfsRecvCmd.Stdin, _ = zfsSendCmd.StdoutPipe()
+		zfsRecvCmd.Stdout = os.Stdout
+		zfsRecvCmd.Stderr = os.Stderr
+
+		err = zfsRecvCmd.Start()
+		if err != nil {
+			return err
+		}
+
+		err = zfsSendCmd.Run()
+		if err != nil {
+			return err
+		}
+
+		err = zfsRecvCmd.Wait()
 		if err != nil {
-			logger.Errorf("Failed to create snapshot for ZFS storage volume \"%s\" on storage pool \"%s\": %s", s.volume.Name, s.pool.Name, err)
 			return err
 		}
 
-		srcDataset := fmt.Sprintf("custom/%s", source.Name)
-		dstDataset := fmt.Sprintf("custom/%s", s.volume.Name)
+		defer zfsPoolVolumeSnapshotDestroy(poolName, fmt.Sprintf("custom/%s", s.volume.Name), tmpSnapshotName)
 
-		// clone snapshot
-		err = zfsPoolVolumeClone(poolName, srcDataset, snapUUID, dstDataset, dstMountPoint)
+		fs := fmt.Sprintf("custom/%s", s.volume.Name)
+		err = zfsPoolVolumeSet(poolName, fs, "canmount", "noauto")
 		if err != nil {
-			logger.Errorf("Failed to create clone for ZFS storage volume \"%s\" on storage pool \"%s\": %s", s.volume.Name, s.pool.Name, err)
 			return err
 		}
 
-		err = zfsPoolVolumeSet(poolName, dstDataset, "mountpoint", dstMountPoint)
+		err = zfsPoolVolumeSet(poolName, fs, "mountpoint", targetStorageVolumeMountPoint)
 		if err != nil {
-			logger.Errorf("Failed to set mountpoint \"%s\" for ZFS storage volume \"%s\" on storage pool \"%s\": %s", dstMountPoint, s.volume.Name, s.pool.Name, err)
 			return err
 		}
+	}
 
-		if !shared.IsMountPoint(dstMountPoint) {
-			zfsMount(poolName, dstDataset)
+	logger.Infof(successMsg)
+	return nil
+}
+
+/*
+func (s *storageZfs) StoragePoolVolumeCopy2(source *api.StorageVolumeSource) error {
```

Comment:
I take it this commented out function can be removed? :)

Reply:
Yes :)